test/e2e/operator_test.go

```diff
@@ -0,0 +1,82 @@
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	// kclient is the Kubernetes client used for e2e tests.
+	kclient client.Client
+	// ctx is an empty context used for client calls.
+	ctx = context.TODO()
+	// operatorName is the name of the operator.
+	operatorName = "contour-operator"
+	// operatorNs is the name of the operator's namespace.
+	operatorNs = "contour-operator"
+	// defaultContourNs is the default spec.namespace.name of a Contour.
+	defaultContourNs = "projectcontour"
+	// testUrl is the url used to test e2e functionality.
+	testUrl = "http://local.projectcontour.io/"
+	// expectedDeploymentConditions are the expected status conditions of a
+	// deployment.
+	expectedDeploymentConditions = []appsv1.DeploymentCondition{
+		{Type: appsv1.DeploymentAvailable, Status: corev1.ConditionTrue},
+	}
+)
+
+func TestMain(m *testing.M) {
+	cl, err := newClient()
+	if err != nil {
+		os.Exit(1)
+	}
+	kclient = cl
+
+	os.Exit(m.Run())
+}
+
+func TestOperatorDeploymentAvailable(t *testing.T) {
+	t.Helper()
+	if err := waitForDeploymentStatusConditions(t, ctx, kclient, 3*time.Minute, operatorName, operatorNs, expectedDeploymentConditions...); err != nil {
+		t.Errorf("failed to observe expected conditions for deployment %s/%s: %v", operatorNs, operatorName, err)
+	}
+}
+
+func TestDefaultContour(t *testing.T) {
+	testName := "test-default-contour"
+	newContour(t, ctx, kclient, testName, operatorNs)
+	defer deleteContour(t, ctx, kclient, 3*time.Minute, testName, operatorNs)
+	// Defer the namespace deletion since contour.spec.namespace.removeOnDeletion
+	// defaults to false.
+	defer deleteNamespace(t, ctx, kclient, 3*time.Minute, defaultContourNs)
```

Comment:
when I tried this this namespace deletion fails because (I think) the default operator deployment was still present

```
=== RUN   TestDefaultContour
    utils.go:214: curl'd url http://local.projectcontour.io/
    utils.go:114: timed out waiting for namespace projectcontour to be deleted: timed out waiting for the condition
    utils.go:81: deleted contour contour-operator/test-default-contour
--- FAIL: TestDefaultContour (233.10s)
...
$ k get deployments -A
NAMESPACE            NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
contour-operator     contour-operator         1/1     1            1           7m22s
```

Reply:
The operator is running in ns `contour-operator` and the test is deleting the operand ns, i.e. `projectcontour`. Did `projectcontour` eventually get deleted? At times I've seen namespace deletion take a while. The instance of Contour, `contour-operator/test-default-contour` should be deleted first and then the namespace should be deleted. Otherwise the operator may be trying to recreate the `projectcontour` namespace. I may remove `defer deleteNamespace()` since it's currently unneeded. It can be reintroduced as additional e2e tests are added.